### DISCUSSION TITLE
How can we run multiple test containers that use mongodb at the same time?

### DISCUSSION THREAD

**Discussion by PatrickGhosn** [testcontainers/testcontainers-dotnet#588] [Q&A] [answered]
Currently, if more than one of our pipelines run on the same machine, they will run into the issue of both trying to use port 27017 to run a mongodb container for the integration tests. Is there a way to get around that limitation and be able to run two or more instances of the same integration tests at the same time? Perhaps by being able to dynamically assign the container port through a shell variable.

**Comment by HofmeisterAn** [comment, marked as answer]
> Is there a way to get around that limitation and be able to run two or more instances of the same integration tests at the same time?

Yes, Testcontainers maps the container port (in your case `27017`) automatic to a random host port. This allows running multiple instances at the same time. It is the default behaviour.

I assume you bind a fix host port. Can you share your configuration?

**Reply by PatrickGhosn** [reply]
Here's the configuration we're using:
``` 
        [AssemblyInitialize]
        public static async Task AssemblyInitialize(TestContext context)
        {
            var containerNamePrefix = "probe-appdb-tests";
            var testcontainersBuilder = new TestcontainersBuilder<TestcontainersContainer>()
              .WithName($"{containerNamePrefix}-mongodb-{Guid.NewGuid()}")
              .WithImage("bitnami/mongodb:4.4.3-debian-10-r21")
              .WithPortBinding(port:27017)
              .WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(27017));

            testcontainers = testcontainersBuilder.Build();

            await testcontainers.StartAsync();
        }
```

I noticed that `WithPortBindings` takes a second parameter `bool assignRandomHostPort = false`, but when I set it to `true`, I get the below error, even when I remove the `UntilPortIsAvailable`:

```
  Message: 
Initialization method Project.DB.MongoTests.AppCtxt.Repositories.AppCtxtRepositoryTests.InitTest threw exception. System.TimeoutException: A timeout occurred after 30000ms selecting a server using CompositeServerSelector{ Selectors = MongoDB.Driver.MongoClient+AreSessionsSupportedServerSelector, LatencyLimitingServerSelector{ AllowedLatencyRange = 00:00:00.0150000 }, OperationsCountServerSelector }. Client view of cluster state is { ClusterId : "1", Type : "Unknown", State : "Disconnected", Servers : [{ ServerId: "{ ClusterId : 1, EndPoint : "Unspecified/localhost:27017" }", EndPoint: "Unspecified/localhost:27017", ReasonChanged: "Heartbeat", State: "Disconnected", ServerVersion: , TopologyVersion: , Type: "Unknown", HeartbeatException: "MongoDB.Driver.MongoConnectionException: An exception occurred while opening a connection to the server.
 ---> System.Net.Sockets.SocketException (10061): No connection could be made because the target machine actively refused it.
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask.ValueTaskSourceAsTask.<>c.<.cctor>b__4_0(Object state)
--- End of stack trace from previous location ---
   at MongoDB.Driver.Core.Connections.TcpStreamFactory.ConnectAsync(Socket socket, EndPoint endPoint, CancellationToken cancellationToken)
   at MongoDB.Driver.Core.Connections.TcpStreamFactory.CreateStreamAsync(EndPoint endPoint, CancellationToken cancellationToken)
   at MongoDB.Driver.Core.Connections.BinaryConnection.OpenHelperAsync(CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at MongoDB.Driver.Core.Connections.BinaryConnection.OpenHelperAsync(CancellationToken cancellationToken)
   at MongoDB.Driver.Core.Servers.ServerMonitor.InitializeConnectionAsync(CancellationToken cancellationToken)
   at MongoDB.Driver.Core.Servers.ServerMonitor.HeartbeatAsync(CancellationToken cancellationToken)", LastHeartbeatTimestamp: "2022-09-21T07:26:23.5996563Z", LastUpdateTimestamp: "2022-09-21T07:26:23.5996566Z" }] }..

  Stack Trace: 
Cluster.ThrowTimeoutException(IServerSelector selector, ClusterDescription description)
WaitForDescriptionChangedHelper.HandleCompletedTask(Task completedTask)
Cluster.WaitForDescriptionChangedAsync(IServerSelector selector, ClusterDescription description, Task descriptionChangedTask, TimeSpan timeout, CancellationToken cancellationToken)
Cluster.SelectServerAsync(IServerSelector selector, CancellationToken cancellationToken)
MongoClient.AreSessionsSupportedAfterServerSelectionAsync(CancellationToken cancellationToken)
MongoClient.AreSessionsSupportedAsync(CancellationToken cancellationToken)
MongoClient.StartImplicitSessionAsync(CancellationToken cancellationToken)
MongoDatabaseImpl.UsingImplicitSessionAsync(Func`2 funcAsync, CancellationToken cancellationToken)
TestHelper.CleanDbAsync(IServiceProvider serviceProvider) line 110
AppCtxtRepositoryTests.InitTest() line 28

```

What's weird is, when I tried to simulate a blocked `27017` host port by using `port-forward` to an existing mongo container, I got an authentication error because it was trying to connect to it and not the new temporary container.

I think it's still connecting to the `27017` host port instead of the random one, even with the parameter set to `true`, but I'm not sure why that's happening.

**Reply by HofmeisterAn** [reply]
Ok, you are not using the pre-configured [MongoDB](https://github.com/testcontainers/testcontainers-dotnet/blob/997c019428132ebc94dc981b454324e1e7a6f302/src/Testcontainers/Configurations/Modules/Databases/MongoDbTestcontainerConfiguration.cs) module. But that is not a problem at all. You are on the right track.

Set `assignRandomHostPort` to `true`. This will map a random host port to the default MongoDB port and allows running multiple instances. Keep the wait strategy. It checks if the port inside the container is available. I suggest a more advanced wait strategy (see the module configuration). Usually it is not enough to just wait until the port is available. To connect to the database use `GetMappedPublicPort(27017)` to receive the random assigned host port e.g. like: https://github.com/testcontainers/testcontainers-dotnet/blob/997c019428132ebc94dc981b454324e1e7a6f302/src/Testcontainers/Containers/Modules/HostedServiceContainer.cs#L26

**Reply by PatrickGhosn** [reply]
Great! Thank you. The issue was that we weren't using the MappedPublicPort to connect. I'll check out the advanced wait strategy since we were having inconsistent initialization in the tests startups and that might resolve it.

**Reply by PatrickGhosn** [reply]
Quick question: I noticed this https://github.com/testcontainers/testcontainers-dotnet/issues/449 has been closed, but we are encoutnering it even after adding the random port. Does this have to do with the wait strategy or something else?

**Reply by HofmeisterAn** [reply]
> Quick question: I noticed this #449 has been closed, but we are encoutnering it even after adding the random port. Does this have to do with the wait strategy or something else?

On Windows (Docker Desktop for Windows)?

**Reply by PatrickGhosn** [reply]
On both WSL (Docker desktop) as well as Linux in Azure Devops. I tried adding delays as well as a wait strategy as you suggested, but it's still happening.

```             
 .WithWaitStrategy(
   Wait.ForUnixContainer()
  .UntilPortIsAvailable(27017)
  .UntilCommandIsCompleted("mongo", $"localhost:27017", "--eval", "db.runCommand(\"ping\").ok", "--quiet")
               )
  .WithStartupCallback((resourceReaper, ct) => Task.Delay(TimeSpan.FromSeconds(15), ct));```

**Reply by HofmeisterAn** [reply]
You are using a private registry, right? Does it contain or proxy the resource reaper image?

**Reply by PatrickGhosn** [reply]
The mongo and reaper images are the default ones, so I think from the public docker registiry. And there are no proxies, outgoing access is direct.

I saw the following suggested to be added in another thread:
```
            await ResourceReaper.GetAndStartDefaultAsync()
                .ConfigureAwait(false);
```
Do you think it might resolve the issue?

**Reply by HofmeisterAn** [reply]
> Do you think it might resolve the issue?

No. I would not recommend it as a final solution. Does it happens frequently? Can you start the default resource reaper standalone? Is Testcontainers able to initialize it?

**Reply by PatrickGhosn** [reply]
It happens around 1-2 times every 5 runs. Usually restarting the pipeline after a fail has it run normally. Usually it happens if there's another pipeline running and is using up the CPU.

Is it possible to keep a resource reaper running constantly on the machine?

**Reply by HofmeisterAn** [reply]
Oh wow, that is way too much and very unusual. If the issue relates to the load, try following:

```csharp
var resourceReaperSessionId = Guid.NewGuid();

_ = await ResourceReaper.GetAndStartNewAsync(resourceReaperSessionId, initTimeout: TimeSpan.FromMinutes(5))
  .ConfigureAwait(false);

_ = new TestcontainersBuilder<TestcontainersContainer>()
  .WithResourceReaperSessionId(resourceReaperSessionId)
  .WithImage(...)
  .Build();
```

Make sure you run `GetAndStartNewAsync` only once and use the session id everywhere.

> Is it possible to keep a resource reaper running constantly on the machine?

No
